### PR TITLE
feat: Support export to postman customize workspace 

### PR DESCRIPTION
**Pull request by gcdd1993**
feature #283

### CODE REVIEW

**Comment by gcdd1993** [comment]
@tangcent ubuntu CI test不知道怎么没通过，看了一下没看明白怎么修改。本地倒是可以运行的

**Comment by tangcent** [comment]
本地执行: `./gradlew check --stacktrace`

---

`PostmanWorkspaceChecker`没有配置默认实现，postman相关的测试用例都无法通过。
要么配`ImplementedBy`,要么在测试用例里用到的地方都手动配一下。

**Comment by gcdd1993** [comment]
> 本地执行: `./gradlew check --stacktrace`
> 
> `PostmanWorkspaceChecker`没有配置默认实现，postman相关的测试用例都无法通过。
> 要么配`ImplementedBy`,要么在测试用例里用到的地方都手动配一下。

本地执行`./gradlew check --stacktrace`成功了，但是`CI check` ...，求助求助
image

**Comment by tangcent** [comment]
`PostmanFormatterTest`里缺少注入:
可以执行: `./gradlew :idea-plugin:test --stacktrace --tests "com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest.testRequest2Item"`

```log
com.google.inject.ConfigurationException: Guice configuration errors:

1) No implementation for com.itangcent.idea.plugin.api.export.postman.PostmanApiHelper was bound.
  while locating com.itangcent.idea.plugin.api.export.postman.PostmanApiHelper
    for field at com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper.postmanApiHelper(PostmanSettingsHelper.kt:21)
  while locating com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper
    for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatter.postmanSettingsHelper(PostmanFormatter.kt:33)
  while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatter
    for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest.postmanFormatter(PostmanFormatterTest.kt:23)
  while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest

2) No implementation for com.itangcent.idea.plugin.settings.helper.PostmanWorkspaceChecker was bound.
  while locating com.itangcent.idea.plugin.settings.helper.PostmanWorkspaceChecker
    for field at com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper.postmanWorkspaceChecker(PostmanSettingsHelper.kt:21)
  while locating com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper
    for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatter.postmanSettingsHelper(PostmanFormatter.kt:33)
  while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatter
    for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest.postmanFormatter(PostmanFormatterTest.kt:23)
  while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest

2 errors
	at com.google.inject.internal.InjectorImpl.getMembersInjector(InjectorImpl.java:1023)
	at com.google.inject.internal.InjectorImpl.getMembersInjector(InjectorImpl.java:1029)
	at com.google.inject.internal.InjectorImpl.injectMembers(InjectorImpl.java:1013)
	at com.itangcent.intellij.context.ActionContext.init(ActionContext.kt:428)
	at com.itangcent.testFramework.ContextLightCodeInsightFixtureTestCase.setUp(ContextLightCodeInsightFixtureTestCase.kt:86)
	at com.intellij.testFramework.UsefulTestCase.defaultRunBare(UsefulTestCase.java:380)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:29)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:27)
	at com.intellij.testFramework.EdtTestUtilKt$runInEdtAndWait$2.run(EdtTestUtil.kt:62)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:346)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by gcdd1993** [comment]
> `PostmanFormatterTest`里缺少注入:
> 可以执行: `./gradlew :idea-plugin:test --stacktrace --tests "com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest.testRequest2Item"`
> 
> ```
> com.google.inject.ConfigurationException: Guice configuration errors:
> 
> 1) No implementation for com.itangcent.idea.plugin.api.export.postman.PostmanApiHelper was bound.
>   while locating com.itangcent.idea.plugin.api.export.postman.PostmanApiHelper
>     for field at com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper.postmanApiHelper(PostmanSettingsHelper.kt:21)
>   while locating com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper
>     for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatter.postmanSettingsHelper(PostmanFormatter.kt:33)
>   while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatter
>     for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest.postmanFormatter(PostmanFormatterTest.kt:23)
>   while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest
> 
> 2) No implementation for com.itangcent.idea.plugin.settings.helper.PostmanWorkspaceChecker was bound.
>   while locating com.itangcent.idea.plugin.settings.helper.PostmanWorkspaceChecker
>     for field at com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper.postmanWorkspaceChecker(PostmanSettingsHelper.kt:21)
>   while locating com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper
>     for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatter.postmanSettingsHelper(PostmanFormatter.kt:33)
>   while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatter
>     for field at com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest.postmanFormatter(PostmanFormatterTest.kt:23)
>   while locating com.itangcent.idea.plugin.api.export.postman.PostmanFormatterTest
> 
> 2 errors
> 	at com.google.inject.internal.InjectorImpl.getMembersInjector(InjectorImpl.java:1023)
> 	at com.google.inject.internal.InjectorImpl.getMembersInjector(InjectorImpl.java:1029)
> 	at com.google.inject.internal.InjectorImpl.injectMembers(InjectorImpl.java:1013)
> 	at com.itangcent.intellij.context.ActionContext.init(ActionContext.kt:428)
> 	at com.itangcent.testFramework.ContextLightCodeInsightFixtureTestCase.setUp(ContextLightCodeInsightFixtureTestCase.kt:86)
> 	at com.intellij.testFramework.UsefulTestCase.defaultRunBare(UsefulTestCase.java:380)
> 	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:29)
> 	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:27)
> 	at com.intellij.testFramework.EdtTestUtilKt$runInEdtAndWait$2.run(EdtTestUtil.kt:62)
> 	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
> 	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
> 	at java.awt.EventQueue.access$500(EventQueue.java:97)
> 	at java.awt.EventQueue$3.run(EventQueue.java:709)
> 	at java.awt.EventQueue$3.run(EventQueue.java:703)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
> 	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
> 	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:346)
> 	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
> 	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
> 	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
> 	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
> 	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
> 	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
> ```

**Comment by tangcent** [comment]
@gcdd1993 以下`Action`也需要注入`PostmanWorkspaceChecker`:
https://github.com/tangcent/easy-yapi/blob/81eb654a8ff1259d5d52fc6392c083c9195cce5c/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/actions/ApiDashBoardAction.kt#L26-L52
https://github.com/tangcent/easy-yapi/blob/81eb654a8ff1259d5d52fc6392c083c9195cce5c/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/suv/SuvApiExporter.kt#L400-L439

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt [hunk 1]
```diff
@@ -63,19 +72,20 @@ open class DefaultPostmanApiHelper : PostmanApiHelper {
...
         if (returnValue.contains("AuthenticationError")
-                && returnValue.contains("Invalid API Key")) {
+            && returnValue.contains("Invalid API Key")
+        ) {
             logger!!.error("Authentication failed!")
             return
         }
         val returnObj = returnValue.asJsonElement()
         val errorName = returnObj
-                .sub("error")
```
Comment (tangcent) [inline idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:72]: 这些格式化的改动最好去掉，会对review产生干扰。
如果需要做格式化、整理import、修复typo等，可以单独提交一个PR

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt [hunk 2]
```diff
@@ -41,6 +59,130 @@ class PostmanSettingsHelper {
...
+        }
+
+        return null
+    }
+
+    private fun selectWorkspace(module: String): String? {
+        val inputTitle = "Workspace"
+        val workspaces = postmanApiHelper.getAllWorkspaces() ?: return null
+        val workspaceMap = workspaces.map { it.name to it.id }.toMap()
+        return messagesHelper.showEditableChooseDialog(
```
Comment (tangcent) [inline idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt:100]: 这里使用`showEditableChooseDialog`有什么考虑吗? 
似乎并没有输入的需求。

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt [hunk 3]
```diff
@@ -41,6 +59,130 @@ class PostmanSettingsHelper {
...
+                    setWorkspace(module, modulePrivateWorkspace!!)
+                    return modulePrivateWorkspace
+                }
+            }
+        }
+
+        return null
+    }
+
+    private fun selectWorkspace(module: String): String? {
```
Comment (tangcent) [inline idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt:96]: 这个方法单元测试没有覆盖到,考虑增加对应的测试用例:
https://github.com/tangcent/easy-yapi/blob/master/idea-plugin/src/test/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelperTest.kt

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt [hunk 4]
```diff
@@ -104,12 +114,18 @@ open class DefaultPostmanApiHelper : PostmanApiHelper {
...
-                .header("x-api-key", postmanSettingsHelper.getPrivateToken())
-                .body(KV.by("collection", collection))
+            .post(COLLECTION)
+            .contentType(ContentType.APPLICATION_JSON)
+            .header("x-api-key", postmanSettingsHelper.getPrivateToken())
+            .body(KV.by("collection", collection))
+
+        if (workspace != null) {
+            request.query("workspace", workspace)
+        }
```
Comment (tangcent) [inline idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:128]:
```suggestion
        workspace?.let { request.query("workspace", it) }
```

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt [hunk 5]
```diff
@@ -272,9 +288,78 @@ open class DefaultPostmanApiHelper : PostmanApiHelper {
...
+                onErrorResponse(response)
+
+                return null
+            }
+        } catch (e: Throwable) {
+            logger!!.traceError("Load workspace $workspaceId failed", e)
+            return null
+        }
+    }
+
```
Comment (tangcent) [inline idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:359]: 这一部分都没有测试用例覆盖

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt [hunk 6]
```diff
@@ -104,12 +114,18 @@ open class DefaultPostmanApiHelper : PostmanApiHelper {
      * @return collection id
      */
     override fun createCollection(collection: HashMap<String, Any?>): HashMap<String, Any?>? {
-
+        // get workspace
+        val module = actionContext.callInReadUI { moduleHelper.findModuleByPath(ActionUtils.findCurrentPath()) }
```
Comment (tangcent) [inline idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:118]: 在这一步获取module无法处理这种情况: [UserCtrl.java#L16-L20](https://github.com/Earth-1610/spring-demo/blob/cce465d2299a4dbab3e77775205a4a7934f97249/cn/springboot-demo/src/main/java/com/itangcent/springboot/demo/controller/UserCtrl.java#L16-L20)
当我尝试导出[spring-demo](https://github.com/Earth-1610/spring-demo/tree/master/cn/springboot-demo/src/main/java/com/itangcent/springboot/demo/controller)或者仅导出[UserCtrl.java](https://github.com/Earth-1610/spring-demo/blob/master/cn/springboot-demo/src/main/java/com/itangcent/springboot/demo/controller/UserCtrl.java)时, 均只提示我选择`springboot-demo`对应的`WorkSpace`,而没有提示我选择`user `对应的`WorkSpace`

---

需要根据**开发者是否有一个项目下的接口导入多个`WorkSpace`的需求**来处理这个问题:

1. 如果没有的话，直接关联`Project`->`WorkSpace`是不是更合理?
2. 如果有的话，就比较麻烦了，需要改造: https://github.com/tangcent/easy-yapi/blob/81eb654a8ff1259d5d52fc6392c083c9195cce5c/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/PostmanFormatter.kt#L511-L521
在对`module`进行分组聚合后，再对它们归属的`workspace`进行分组聚合

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt (hunk 1 above)
Comment (gcdd1993) [reply idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:72]: 抱歉，我来恢复一下，IDEA自动给我格式化了

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt (hunk 6 above)
Comment (gcdd1993) [reply idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:118]: 我觉得直接关联`Project` -> `WorkSpace`更为合理。
是我理解错了，我原先的想法就是一个`Project`导入一个`WorkSpace`，错把`module`当成`project`了

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt (hunk 2 above)
Comment (gcdd1993) [reply idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt:100]: 这里并没有特殊的考虑，只是发现`Message#showChooseDialog`标记为**已过期**，而适用的下拉选好像只有这一个
image

**Comment by gcdd1993** [comment]
@tangcent `ApiDashboard`这样改一下是否会更好一点？
如果用户为本项目配置过workspace，则选中该workspace，如果没有，则不选中任何workspace
image

**Comment by tangcent** [comment]
> @tangcent `ApiDashboard`这样改一下是否会更好一点？
> 如果用户为本项目配置过workspace，则选中该workspace，如果没有，则不选中任何workspace
> image
> 

LGTM

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt (hunk 5 above)
Comment (gcdd1993) [reply idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:356]: 这块`PostmanApiHelper`先前也没有对应的测试用例，而且要写单元测试，势必需要暴露`PostmanApiKey`才行

Comment (tangcent) [reply idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt:356]: 那就先不写，我后面有空了来补。
不暴露`PostmanApiKey `,直接把请求都mock掉

**Comment by gcdd1993** [comment]
1、已完成并自测`ApiDashboard`选择`workspace`并导入`collection`
Snipaste_2021-07-13_11-25-57
2、快捷键导出至自定义`workspace`

File: idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt (hunk 3 above)
Comment (gcdd1993) [reply idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/settings/helper/PostmanSettingsHelper.kt]: @tangcent `selectWorkspace`是`private`的方法，这个有什么办法可以写单元测试吗？

**Comment by tangcent** [comment]
@gcdd1993 先合进来再改?

**Comment by gcdd1993** [comment]
> @gcdd1993 先合进来再改?

额，还需要改啥，上面的都改好了

**Pull request merged**